test/reducer-e2e/reducer_test.go

```diff
@@ -0,0 +1,77 @@
+package reducer_e2e
+
+import (
+	dfv1 "github.com/numaproj/numaflow/pkg/apis/numaflow/v1alpha1"
+	. "github.com/numaproj/numaflow/test/fixtures"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+)
+
+type ReducerSuite struct {
+	E2ESuite
+}
+
+func (s *ReducerSuite) TestSimpleReducerPipelineFailOver() {
+	w := s.Given().Pipeline("@testdata/simple-reduce-pipeline.yaml").
+		When().
+		CreatePipelineAndWait()
+	defer w.DeletePipelineAndWait()
+
+	pipelineName := "even-odd-sum"
+
+	w.Expect().
+		VertexPodsRunning().DaemonPodsRunning().
+		VertexPodLogContains("in", LogSourceVertexStarted).
+		VertexPodLogContains("atoi", LogUDFVertexStarted, PodLogCheckOptionWithContainer("numa")).
+		VertexPodLogContains("sink", LogSinkVertexStarted).
+		DaemonPodLogContains(pipelineName, LogDaemonStarted)
+
+	defer w.VertexPodPortForward("in", 8443, dfv1.VertexHTTPSPort).
+		TerminateAllPodPortForwards()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		startTime := int(time.Unix(1000, 0).UnixMilli())
+		for i := 1; i <= 100; i++ {
+			for j := 0; j < 1; j++ {
+				eventTime := startTime + (i * 1000)
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("1")).
+					Expect().
+					Status(204)
+
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("2")).
+					Expect().
+					Status(204)
+
+				HTTPExpect(s.T(), "https://localhost:8443").POST("/vertices/in").WithHeader("X-Numaflow-Event-Time", strconv.Itoa(eventTime)).WithBytes([]byte("3")).
+					Expect().
+					Status(204)
+			}
+		}
+
+		w.Expect().VertexPodLogContains("sink", "Payload -  120")
+		w.Expect().VertexPodLogContains("sink", "Payload -  240")
+	}()
+
+	// Kill one of the pod during processing to trigger failover.
+	block := func(t *testing.T, output string, err error) {
```

Comment:
I am also thinking about whether this should be a utility function that can be used in other tests too.

Reply:
Yeah, I guess only thing may be depends on the scenarios, one may want to take different actions towards the value of `output` or `err`

Reply:
such random testing shouldn't be done in e2e tests. E2E tests should be predictable. Such random tests can be labeled, perhaps under Fuzz tests or something.

There are three major of pod migration scenarios
* right before processing
* during processing
* right after processing

Reply:
are we interpreting the result of `output` or `err` on pod kill?

Reply:
Yes, that is the expected behavior of `deleted` right? Hence I am suggesting we make it reusable. I can't think of any other checks, do you envision any?